### DISCUSSION TITLE
Velero data store in GCP bucket

### DISCUSSION THREAD

**Discussion by au-fetch** [vmware-tanzu/velero#8520] [Community support Q&A] [answered]
I am currently using Velero version 1.14.0 and would like to store my data in a GCP bucket instead of using snapshots.

Is this possible? If so, could you please explain how to achieve it?

**Comment by kaovilai** [comment]
I assume you mean disk PV/PVC data because other parts of velero are already using buckets.

You can opt to use [Data Mover](https://velero.io/docs/v1.15/csi-snapshot-data-movement/) or [File System Backup](https://velero.io/docs/v1.15/file-system-backup/). Please refer to each installation instructions and usage example.

**Reply by Mr-Philipp** [reply]
i am facing the same issue right know, are you sure this is supported?

my installation parameters are:

```
    --provider gcp \
    --plugins velero/velero-plugin-for-gcp:v1.12.0 \ 
    --bucket $BUCKET \
    --secret-file ./credentials-velero \
    --use-node-agent \
    --features=EnableCSI,DataMover
```
and i do backups by calling

```
velero backup create test --snapshot-move-data
```
but volumes and data is not exported into the bucket.

**Reply by blackpiglet** [reply]
The Data mover feature is not enabled like that.

``` bash
velero backup create NAME --snapshot-move-data OPTIONS...
```

Furthermore, you need to enable node-agent and EnableCSI during Velero installation to make it work.

**Reply by Mr-Philipp** [reply]
> The Data mover feature is not enabled like that.
> 
> ```shell
> velero backup create NAME --snapshot-move-data OPTIONS...
> ```
> 
> Furthermore, you need to enable node-agent and EnableCSI during Velero installation to make it work.

Thanks for your lightning fast response @blackpiglet 

i am just a bit confused about your answer because thats exactly what i did, didn't i? 
you said that 

```
velero backup create test --snapshot-move-data
```
is not the way data mover feature is enabled and your provided solution is exactly the same as mine?

Also you said that node-agent and EnableCSI needs to be enabled during installation what i also did. Can you please provide an installation command which enables everything needed to do snapshot exports on GCP because mine seems to be wrong but i dont get what exactly is wrong.

**Reply by blackpiglet** [reply]
I missed the backup command's parameters.
I just wanted to point out this `--features=EnableCSI,DataMover`. The `DataMover` part is useless.

If it works as expected, there should be a new directory called `kopia` in the object storage.
The volume data are stored there. The data are organized by kopia, so the data cannot be read directly.

If there is the `kopia` directory there, please take a restore to see whether the data is correctly restored.

**Reply by Mr-Philipp** [reply]
Thank you for your clarification. I got it working so that it creates a kopia directoy in the bucket with data in it. I need to try the restore tough. Is this a file based export or a snapshot blob export?

I am still a bit confused about all the different moving parts in velero because for GCP it seems like i can trigger the snapshot with the GCP plugin only like in this installation:

```
velero install \
    --provider gcp \
    --plugins velero/velero-plugin-for-gcp:v1.12.0 \
    --bucket $BUCKET \
    --secret-file ./credentials-velero \
```

or i can use the CSI plugin instead (?)

```
velero install \
    --provider gcp \
    --plugins velero/velero-plugin-for-gcp:v1.12.0 \ 
    --bucket $BUCKET \
    --secret-file ./credentials-velero \
    --use-node-agent \
    --features=EnableCSI
```

so if the `DataMover` Feature is useless (why is it useless?), when do i need the `--use-node-agent` or the `--use-volume-snapshots=true` or `--use-volume-snapshots=false` switch? 

and for creating backups do i need to use `--snapshot-move-data` or not?

also i tried `--default-volumes-to-fs-backup` which seems to do a file-based backup and also creates the kopia directory. is it then any different then the "snapshot-export kopia" export or is it both file based?

What i would like to use is a snapshot export (not file based) to objcet storage with auto-cleanup of the google disk snapshots. I am still not sure what EXACTLY would be the optimal setup.

**Reply by kaovilai** [reply]
> so if the DataMover Feature is useless

We only mean that `DataMover` is not a valid parameter of `--features=`. Whichever documentation you got that from needs to be corrected.

**Reply by kaovilai** [reply]
Please follow https://velero.io/docs/v1.16/csi-snapshot-data-movement/ exactly. It outlines the `velero install` requirements and flags to use during `velero backup create`

**Reply by Mr-Philipp** [reply, marked as answer]
thank you for the response. It's a bit confusing but i got it working now. for everyone else wondering: `DataMover` is not a valid parameteras @kaovilai already mentioned. Also `EnableCSI`, `use-node-agent` as well as the gcp plugin is needed.
